tools/src/cmd_pipeline/cmd_show_html.rs

```diff
@@ -0,0 +1,149 @@
+/// then filter lines based on "data-symbols".  It's not immediately clear if
+/// this command should also handle that or not.  There's also the question of
+/// whether symbol filtering would only excerpt the span associated with the
+/// symbol or would walk up to the enclosing line.  The answers should probably
+/// be driven by command-line use-cases; in particular, the experience of
+/// evolving a more targeted query.  Having to modify a command up-stream should
+/// be considered undesirable.
+pub struct ShowHtmlCommand {
+    pub args: ShowHtml,
+}
+
+#[async_trait]
+impl PipelineCommand for ShowHtmlCommand {
+    async fn execute(
+        &self,
+        server: &Box<dyn AbstractServer + Send + Sync>,
+        input: PipelineValues,
+    ) -> Result<PipelineValues> {
+        let jr = match input {
+            PipelineValues::JsonRecords(jr) => jr,
+            _ => JsonRecords { by_file: vec![] },
+        };
+
+        let mut html_by_file = vec![];
+
+        for fr in jr.by_file {
+            // ## For each file!
+            let lines_to_show = fr.line_set();
+
+            // XXX Doing this as a single string received in a lump is fine for
+            // our testing use-case, but this may need to be reconsidered in
+            // production.  Or maybe production really wants the performance?
+            // Production certainly should have the RAM for our known worst
+            // case scenarios.
+            let html_str = server.fetch_html(&fr.file).await?;
+
+            let mut file_excerpts = vec![];
+
+            // ### HTML Extraction: What We Want
+            //
+            // We want the full line container which looks like:
+            // - div id="line-N" class="source-line-with-number" role="row"
+            //   - div role="cell"
+            //     - div class="cov-strip cov-uncovered cov-known"
+            //   - div role="cell"
+            //     - div class="blame-strip c2" data-blame="..."
+            //   - div role="cell" class="line-number" data-line-number="N"
+            //   - code role="cell" class="source-line"
+            //     - ex: span class="sync_comment"
+            //     - ex: span class="syn_def syn_type" data-symbols="..." data-i
+            //
+            // ### HTML Extraction Low Level Details
+            //
+            // Until https://github.com/cloudflare/lol-html/issues/40 or
+            // the spin-off https://github.com/cloudflare/lol-html/issues/78
+            // are implemented, lol_html doesn't explicitly provide a way to
+            // derive the value of an element.
+            //
+            // So we attempt a hack where we use a custom output sink that is
+            // kept aware of where we are in the file.  The good news is that
+            // since lol_html is oriented around minimal memory allocation, we
+            // can generally control when flushes happen.
+
+            let mut writing_line: u32 = 0;
```

Comment:
I didn't look at this code in detail (need to sleep soon) but it might be nice to extract helper functions with a couple of standalone tests. It looks like the kind of code where things can go wrong easily.

Reply:
Ok, sounds good.